core/unmarshal_v2.go

```diff
@@ -0,0 +1,521 @@
+	
+	// Now delegate the work to the appropriate internal unmarshal function.
+	switch rResultType.Kind() {
+	case reflect.Ptr, reflect.Interface:
+		// Unmarshal a single instance of a model.
+		err = unmarshalModelInstance(rawInput, propertyName, result, unmarshaller)
+	case reflect.Slice:
+		// Unmarshal a slice of model instances.
+		err = unmarshalModelSlice(rawInput, propertyName, result, unmarshaller)
+	case reflect.Map:
+		// For a map, we need to look at the map entry type.
+		rEntryType := reflect.TypeOf(result).Elem().Elem()
+		switch rEntryType.Kind() {
+		case reflect.Struct, reflect.Interface:
+			// A map[string]<model>
+			err = unmarshalModelMap(rawInput, propertyName, result, unmarshaller)
+		case reflect.Slice:
+			// A map[string][]<model>
+			err = unmarshalModelSliceMap(rawInput, propertyName, result, unmarshaller)
+		default:
+			err = fmt.Errorf(errorUnsupportedMapEntryType, reflect.TypeOf(result).Elem().Elem().String())
+			return
+		}
+	default:
+		err = fmt.Errorf(errorUnsupportedResultType, reflect.TypeOf(result).Elem().String())
+		return
+	}
+	return
+}
+
+//
+// unmarshalModelInstance unmarshals 'rawInput' into an instance of a model.
+//
+// Parameters:
+//
+// rawInput: the unmarshal input source.
+//
+// propertyName: the name of the property within 'rawInput' that contains the instance of the model.
+// If 'propertyName' is specified as "" then 'rawInput' is assumed to contain the model instance directly.
+//
+// result: should be a ptr to a ptr to a model instance (e.g. **Foo for model type Foo).
+// A new instance of the model will be constructed and returned through the model pointer
+// that 'result' points to (i.e. it is legal for 'result' to point to a nil pointer).
+//
+// 'unmarshaller' is the generated unmarshal function used to unmarshal a single instance of the model.
+//
+func unmarshalModelInstance(rawInput interface{}, propertyName string, result interface{}, unmarshaller ModelUnmarshaller) (err error) {
+	var rawMap map[string]json.RawMessage
+	var ok bool
+
+	// rawInput should be a map[string]json.RawMessage
+	rawMap, ok = rawInput.(map[string]json.RawMessage)
+	if !ok {
+		err = fmt.Errorf(errorIncorrectInputType, "map[string]json.RawMessage", reflect.TypeOf(rawInput).String())
+		return
+	}
+
+	// If propertyName was specified, then retrieve that entry from 'rawInput' as our unmarshal input source.
+	// Otherwise, just use 'rawInput' directly.
+	if propertyName != "" {
+		rawMsg, foundIt := rawMap[propertyName]
+		if foundIt && rawMsg != nil {
+			// rawMsg should contain an instance of the model, so we need to unmarshal it into
+			// a map[string]RawMessage that we can pass to 'unmarshaller'.
+			rawMap = make(map[string]json.RawMessage)
+			err = json.Unmarshal(rawMsg, &rawMap)
+			if err != nil {
+				err = fmt.Errorf(errorUnmarshalModel, propInsert(propertyName),
+					reflect.TypeOf(result).Elem().Elem().String(), err.Error())
+				return
+			}
+		} else {
+			rawMap = nil
```

Comment:
if ``rawMap`` is nil this method will return no error, but what is the expected unmarshal result? Will ``result`` still be an empty interface?

Reply:
In the case where the specified property is not found in the input map, we want to essentially return a nil model pointer.
That's why we initialize te result to nil, then do the actual unmarshalling only if rawMap is not nil.